### PR TITLE
An initial design for multiple IaaS implementations in IBM Cloud

### LINKED ISSUE #302: Design for general ibm cloud infrastructure against specific IaaS type
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
In IBM Cloud we have different types of Iaas, VPC gen2, PowerVS, Z etc.
We should make sure this provider can cover different IBM Cloud IaaS respecting the same pattern rather than many specific implementations.

### PR DESCRIPTION
**Pull request by wentao-zh**
**What this PR does / why we need it**:
To support different types of IaaS implementaion in IBM Cloud we need to organize the specific IBM IaaS Object very carefully. To reduce the complexity and provide flexibility.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #302 

**Specify your PR reviewers**:

/cc @gyliu513 @jichenjc @xunpan

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by gyliu513** [comment]
Thanks @wentao-zh , looks good. Can you help check why CI failed, and another is can you help add toc via [doctoc](https://github.com/thlorenz/doctoc)?

**Review by mkumatag — APPROVED** [review]

File: docs/proposal/architecture_for_diff_iaas.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+## Motivation
+
+There are multiple types of IaaS in IBM Cloud: Classic infrastructure, VPC and PowerVS etc. Each type of IaaS has some specific features. Users would like to bootstrap a k8s cluster from any types of IaaS. It's hard to maintain an individual repository for each type of IaaS. A good implementation should describe IBM Cloud feature in an appropriate grain. 
+
+## Proposal
+
+This proposal presents a 2 layers infrastructure resource definition. Th top layer describes IBM Cloud common feature and the 2nd layer describes the details of the specific IaaS implementation.
+
+![](../images/cluster-api-ibmcloud-definition.png)
+
```
Comment (jichenjc) [inline docs/proposal/architecture_for_diff_iaas.md:12]: I assume one cluster-api-ibmcloud pod will watch all of the resources and in one cluster we can handle all those 3 resources?

File: docs/proposal/architecture_for_diff_iaas.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+If user need to bootstrap k8s from PowerVS. The implementation is similar.
+
+![](../images/cluster-api-ibmcloud-powervs.png)
+
+## Benefit of this design
+
+### Unifined IaaS resource definition for cluster-api
+
+Actually the top level Object is things like cluster-api object in IBM Cloud. Cluster api intends to provide an abstract definition for different types of cloud-providers. IBMCluster intends to provide abstract definition for different types of IaaS implementation in IBM Cloud. Owner reference and infrastructureRef relation ship can be established as cluster-api object and provider object.
```
Comment (jichenjc) [inline docs/proposal/architecture_for_diff_iaas.md:28]: s/relation ship/ relationship/

File: docs/proposal/architecture_for_diff_iaas.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+## Proposal
+
+This proposal presents a 2 layers infrastructure resource definition. Th top layer describes IBM Cloud common feature and the 2nd layer describes the details of the specific IaaS implementation.
+
+![](../images/cluster-api-ibmcloud-definition.png)
+
+IBMCluster/IBMMachine/IBMMachineTemplate is the generic top layer. It would only show the required information to reconsile status with cluster-api level Objects. Cluster, Machine and MachineDeployment etc. It doesn't know that VPC infrastructure would care about subnet details, public gateways, security groups. And it also doesn't care about PowerVS needs CloudInstance.
+
+We borrowed the `infrastructureRef` concept from cluster-api level Object. An IBMCluster should reference a specific IaaS implementation. Using VPC as an example we have an individual `IBMVPCCluster` object which has all the details that are needed in IBM Cloud VPC. There is also a specific ibmvpccluster_controller to watch the `IBMVPCCluster` and manuplate the VPC, subnet, security group and public gateway. They are all VPC resouces.
```
Comment (jichenjc) [inline docs/proposal/architecture_for_diff_iaas.md:15]: I understand `infrastructureRef` will be used , it should be good to go this way


in the picture below ,who is responsible for define `IBMPowerVSCluster` and `IBMPowerVSMachine`?

File: docs/proposal/architecture_for_diff_iaas.md (hunk 1 above)
Comment (gyliu513) [reply docs/proposal/architecture_for_diff_iaas.md:12]: Yes

File: docs/proposal/architecture_for_diff_iaas.md (hunk 3 above)
Comment (gyliu513) [reply docs/proposal/architecture_for_diff_iaas.md:15]: Should be PowerVS team

File: docs/proposal/architecture_for_diff_iaas.md (hunk 2 above)
Comment (gyliu513) [reply docs/proposal/architecture_for_diff_iaas.md:28]: @wentao-zh can you update?

File: docs/proposal/architecture_for_diff_iaas.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+There are multiple types of IaaS in IBM Cloud: Classic infrastructure, VPC and PowerVS etc. Each type of IaaS has some specific features. Users would like to bootstrap a k8s cluster from any types of IaaS. It's hard to maintain an individual repository for each type of IaaS. A good implementation should describe IBM Cloud feature in an appropriate grain. 
+
+## Proposal
+
+This proposal presents a 2 layers infrastructure resource definition. Th top layer describes IBM Cloud common feature and the 2nd layer describes the details of the specific IaaS implementation.
+
+![](../images/cluster-api-ibmcloud-definition.png)
+
+IBMCluster/IBMMachine/IBMMachineTemplate is the generic top layer. It would only show the required information to reconsile status with cluster-api level Objects. Cluster, Machine and MachineDeployment etc. It doesn't know that VPC infrastructure would care about subnet details, public gateways, security groups. And it also doesn't care about PowerVS needs CloudInstance.
+
```
Comment (spzala) [inline docs/proposal/architecture_for_diff_iaas.md:14]: manipulate (typo - manuplate)

File: docs/proposal/architecture_for_diff_iaas.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+We borrowed the `infrastructureRef` concept from cluster-api level Object. An IBMCluster should reference a specific IaaS implementation. Using VPC as an example we have an individual `IBMVPCCluster` object which has all the details that are needed in IBM Cloud VPC. There is also a specific ibmvpccluster_controller to watch the `IBMVPCCluster` and manuplate the VPC, subnet, security group and public gateway. They are all VPC resouces.
+Similarly `IBMVPCMachine` stands for compute instance in VPC scope. You can specify image id, ssh key and subnet for the compute instance which would be used as control plane or worker nodes. The ibmvpcmachine_controller can manuplate the compute in VPC world. Both controllers work through vpc sdk.
+
+![](../images/cluster-api-ibmcloud-vpc.png)
+
+If user need to bootstrap k8s from PowerVS. The implementation is similar.
+
+![](../images/cluster-api-ibmcloud-powervs.png)
+
+## Benefit of this design
```
Comment (spzala) [inline docs/proposal/architecture_for_diff_iaas.md:24]: nit - Benefits instead of Benefit

File: docs/proposal/architecture_for_diff_iaas.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+![](../images/cluster-api-ibmcloud-definition.png)
+
+IBMCluster/IBMMachine/IBMMachineTemplate is the generic top layer. It would only show the required information to reconsile status with cluster-api level Objects. Cluster, Machine and MachineDeployment etc. It doesn't know that VPC infrastructure would care about subnet details, public gateways, security groups. And it also doesn't care about PowerVS needs CloudInstance.
+
+We borrowed the `infrastructureRef` concept from cluster-api level Object. An IBMCluster should reference a specific IaaS implementation. Using VPC as an example we have an individual `IBMVPCCluster` object which has all the details that are needed in IBM Cloud VPC. There is also a specific ibmvpccluster_controller to watch the `IBMVPCCluster` and manuplate the VPC, subnet, security group and public gateway. They are all VPC resouces.
+Similarly `IBMVPCMachine` stands for compute instance in VPC scope. You can specify image id, ssh key and subnet for the compute instance which would be used as control plane or worker nodes. The ibmvpcmachine_controller can manuplate the compute in VPC world. Both controllers work through vpc sdk.
```
Comment (spzala) [inline docs/proposal/architecture_for_diff_iaas.md:16]: manipulate (typo - manuplate)

**Review by spzala — COMMENTED** [review]
Great, thanks @wentao-zh I just had couple of nits.

**Pull request merged**